hxuncdata.1
.TH "HXUNCDATA" "1" "10 Jul 2011" "7.x" "HTML-XML-utils"
.SH NAME
hxuncdata \- replace CDATA sections by character entities
.SH SYNOPSIS
.B hxuncdata
.RI "[\| " xml-file " \|]"
.SH DESCRIPTION
.B hxuncdata
replaces CDATA sections by character entities. The input is scanned
for occurrences of "<![CDATA[" and corresponding "]]>". Those strings
are removed and all occurrences of "&", "<" and ">" in between them
will be replaced by "&amp;", "&lt;" and "&gt;" respectively.
.PP
The input must be 1 byte per character. If it is not, convert it to
UTF-8 first.
.SH OPTIONS
The following options are supported:
.TP
.BR \-? ", " \-\-help
Show command usage.
.SH OPERANDS
The following operand is supported:
.TP
.I xml-file
The file to work on.
.SH SEE ALSO
.BR UTF-8 " (RFC 2279)"


ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash
	escapes also outside ASCII range (such characters are converted to
	UTF-8). Bug fixes.

2016-10-17  Bert Bos  <bert@w3.org>

	* Released version 7.1.


hxuncdata.c
/*
 * uncdata -- remove CDATA sections from an XML file
 *
 * The input is scanned for occurrences of "<![CDATA[" and
 * corresponding "]]>". Those strings are removed and all occurrences
 * of "&", "<" and ">" in between them will be replaced by "&amp;",
 * "&;t;" and "&gt;" resp.
 *
 * The input must be 1 byte per character. If it is not, convert it to
 * UTF-8 first.
 *
 * Part of HTML-XML-utils, see:
 * http://www.w3.org/Tools/HTML-XML-utils/
 *
 * Author: Bert Bos <bert@w3.org>
 * Created: 20 Feb 2002
 * Version: $Id: hxuncdata.c,v 1.2 2009/01/08 14:35:09 bbos Exp $
 */

#include <stdio.h>
#include <assert.h>
#include <string.h>


/* process -- process one file */
static void process(FILE *f)
{
  int c;
  enum {INITIAL, START, CDATA1, CDATA2, CDATA3, CDATA4, CDATA5,
	CDATA6, CDATA7, CDATA98, CDATA99, CDATA, MARKUP, DECL1, DECL,
	COMMENT1, COMMENT, COMMENT99, DQUOTE, SQUOTE} state = INITIAL;

  /* No attempt at reporting errors for impossible XML,
     and no support for internal DTD subsets
  */
  while ((c = getc(f)) != EOF) {
    switch (state) {
    case INITIAL:
      if (c == '<') state = START;
      else putchar(c);
      break;
    case START:					/* Seen "<" */
      if (c == '!') state = DECL1;
      else if (c == '>') {putchar('<'); putchar('>'); state = INITIAL;}
      else {putchar('<'); putchar(c); state = MARKUP;}
      break;
    case MARKUP:				/* Inside "<...>" */
      if (c == '"') {putchar('"'); state = DQUOTE;}
      else if (c == '\'') {putchar('\''); state = SQUOTE;}
      else if (c == '>') {putchar('>'); state = INITIAL;}
      else putchar(c);
      break;
    case DQUOTE:				/* Inside double quotes */
      if (c == '"') {putchar('"'); state = MARKUP;}
      else putchar(c);
      break;
    case SQUOTE:				/* Inside single quotes */
      if (c == '\'') {putchar('\''); state = MARKUP;}
      else putchar(c);
      break;
    case DECL1:					/* Seen "<!" */
      if (c == '-') {printf("<!-"); state = COMMENT1;}
      else if (c == '[') state = CDATA1;
      else {putchar('<'); putchar('!'); putchar(c); state = DECL;}
      break;
    case DECL:					/* Inside "<!...>" */
      if (c == '-') {putchar('-'); state = COMMENT1;}
      else if (c == '>') {putchar('>'); state = INITIAL;}
      else putchar(c);
      break;
    case COMMENT1:				/* Seen "-" */
      if (c == '-') {putchar('-'); state = COMMENT;}
      else {putchar(c); state = DECL;}
      break;
    case COMMENT:				/* Seen "--" */
      if (c == '-') {putchar('-'); state = COMMENT99;}
      else putchar(c);
      break;
    case COMMENT99:				/* Seen "-" */
      if (c == '-') {putchar('-'); state = DECL;}
      else {putchar(c); state = COMMENT;}
      break;
    case CDATA1:				/* Seen "<![" */
      if (c == 'C') state = CDATA2;
      else {printf("<![%c", c); state = INITIAL;}
      break;
    case CDATA2:				/* Seen "<![C" */
      if (c == 'D') state = CDATA3;
      else {printf("<![C%c", c); state = INITIAL;}
      break;
    case CDATA3:				/* Seen "<![CD" */
      if (c == 'A') state = CDATA4;
      else {printf("<![CD%c", c); state = INITIAL;}
      break;
    case CDATA4:				/* Seen "<![CDA" */
      if (c == 'T') state = CDATA5;
      else {printf("<![CDA%c", c); state = INITIAL;}
      break;
    case CDATA5:				/* Seen "<![CDAT" */
      if (c == 'A') state = CDATA6;
      else {printf("<![CDAT%c", c); state = INITIAL;}
      break;
    case CDATA6:				/* Seen "<![CDATA" */
      if (c == '[') state = CDATA;
      else {printf("<![CDATA%c", c); state = INITIAL;}
      break;
    case CDATA:					/* Inside "<![CDATA[...]]>" */
      if (c == ']') state = CDATA98;
      else if (c == '<') fputs("&lt;", stdout);
      else if (c == '>') fputs("&gt;", stdout);
      else if (c == '&') fputs("&amp;", stdout); 
      else putchar(c);
      break;
    case CDATA98:				/* Seen "]" */
      if (c == ']') state = CDATA99;
      else {putchar(']'); putchar(c); state = CDATA;}
      break;
    case CDATA99:				/* Seen "]]" */
      if (c == '>') state = INITIAL;
      else {putchar(']'); putchar(']'); putchar(c); state = CDATA;}
      break;
    default:
      assert(!"Cannot happen!");
    }
  }
}

int main(int argc, char *argv[])
{
  int i, err = 0;
  FILE *f;

  if (argc == 1)
    process(stdin);
  else if (strcmp(argv[1], "--help") == 0 || strcmp(argv[1], "-?") == 0)
    printf("Usage: %s [XML-FILE]...\n", argv[0]);
  else
    for (i = 1; i < argc; i++) {
      if (!(f = fopen(argv[i], "r"))) {perror(argv[i]); err++; continue;}
      process(f);
      if (fclose(f) != 0) {perror(argv[i]); err++; continue;}
    }
  return err;
}
